### PR TITLE
Clear softmax_cross_entropy deprecation warning.

### PR DESCRIPTION
**Pull request by MarkDaoust**
The `sparse` version is more efficient anyway.

I'm returning the labels shape [1] instead of []
because tf.accuracy fails otherwise.

mnist_test.py passes.
I manually ran both `mnist.py` and `mnist_tpu.py` with a reduced dataset to confirm they haven't broken .

### CODE REVIEW

File: official/mnist/dataset.py [hunk 1]
```diff
@@ -89,15 +89,14 @@ def decode_image(image):
     image = tf.reshape(image, [784])
     return image / 255.0
 
-  def one_hot_label(label):
-    label = tf.decode_raw(label, tf.uint8)  # tf.string -> tf.uint8
-    label = tf.reshape(label, [])  # label is a scalar
-    return tf.one_hot(label, 10)
+  def decode_label(label):
+    label = tf.decode_raw(label, tf.uint8)  # tf.string -> [tf.uint8]
+    return tf.to_int32(label)
```
Comment (asimshankar) [inline official/mnist/dataset.py]: Random thought: Would it make sense for the dataset to return a `uint8` tensor and push the casting to the point where it is required?  (The answer may well be no :))

**Comment by nealwu** [comment]
Jenkins, test this please.

**Comment by nealwu** [comment]
Jenkins, test this please.

**Comment by nealwu** [comment]
Jenkins, test this please.

File: official/mnist/dataset.py (hunk 1 above)
Comment (MarkDaoust) [reply official/mnist/dataset.py]: It's not big difference either way.

I add a commit to move the cast to the top of the model_fns.

**Comment by nealwu** [comment]
Jenkins, test this please.

**Comment by nealwu** [comment]
Jenkins, test this please.

**Comment by nealwu** [comment]
Hey @MarkDaoust, looks like there are some test failures. I just verified that tests pass with the current master code: https://github.com/tensorflow/models/pull/3238.

Could you take a look and fix the failures?

**Comment by MarkDaoust** [comment]
Jenkins, test this please.

**Comment by MarkDaoust** [comment]
Rebased on master without the labels-casting change.

**Comment by nealwu** [comment]
Jenkins, test this please.

**Comment by nealwu** [comment]
Jenkins, test this please.

**Comment by nealwu** [comment]
Hey @MarkDaoust, this PR resulted in MNIST not converging at all. When I train it now I end up at 9.9% accuracy which is a bit worse than random guessing. Going to revert it for now until we can figure out what happened.

**Comment by MarkDaoust** [comment]
Yeah, if it's not working roll it back.

If I checkout the commit before the rollback (`6ab25864f01cbcc6ef508`), it seems to be converging:

```None
python mnist.py
```

``` None
INFO:tensorflow:loss = 2.3103716, step = 1
INFO:tensorflow:global_step/sec: 5.4052
INFO:tensorflow:train_accuracy = 0.47265625 (18.501 sec)
INFO:tensorflow:loss = 0.47470608, step = 101 (18.501 sec)
INFO:tensorflow:global_step/sec: 5.18013
INFO:tensorflow:train_accuracy = 0.6145833 (19.304 sec)
INFO:tensorflow:loss = 0.31265426, step = 201 (19.304 sec)
INFO:tensorflow:global_step/sec: 5.00004
INFO:tensorflow:train_accuracy = 0.69921875 (20.000 sec)
INFO:tensorflow:loss = 0.16615105, step = 301 (20.000 sec)
```

And the mnist_tpu run locally:

```
...
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
INFO:tensorflow:Saving checkpoints for 1 into /var/folders/0s/5q9kfzfj3gx2knj0vj8p68yc00dhcr/T/tmpd04vWs/model.ckpt.
INFO:tensorflow:loss = 2.2985544, step = 0
INFO:tensorflow:global_step/sec: 4.54734
INFO:tensorflow:loss = 0.40759182, step = 100 (21.992 sec)
INFO:tensorflow:Saving checkpoints for 200 into /var/folders/0s/5q9kfzfj3gx2knj0vj8p68yc00dhcr/T/tmpd04vWs/model.ckpt.
INFO:tensorflow:Loss for final step: 0.29425052.
...
INFO:tensorflow:Saving dict for global step 200: accuracy = 0.928125, global_step = 200, loss = 0.246332
```
It's converging for me, so I'm not sure what's up.

**Comment by nealwu** [comment]
I also get increasing train accuracy, but it's the validation accuracy that never improves for me.
